numpy/lib/function_base.py

```diff
@@ -4014,122 +4116,233 @@ def _quantile_is_valid(q):
+    # We need at least two values to do an interpolation
+    too_few_values = valid_values_count < 2
+    if too_few_values.any():
+        valid_values_count[too_few_values] = np.NaN
+    # --- Computation of indexes
+    # Add axis for quantiles
+    valid_values_count = valid_values_count[..., np.newaxis]
+    # Index where to find the value in the sorted array.
+    # Virtual because it is a floating point value, not an valid index.
+    # The nearest neighbours are used for interpolation
+    virtual_indexes = np.where(
+        valid_values_count == 0,
+        0.0,
+        interpolation.virtual_index_fun(valid_values_count, quantiles),
+    )
+    # After sort, slices having NaNs will have for last element a NaN
+    previous_indexes = np.floor(virtual_indexes)
+    next_indexes = previous_indexes + 1
+    # When indexes is an integer there is no interpolation to do
+    integer_indexes = virtual_indexes % 1 == 0
+    if integer_indexes.any():
+        next_indexes[integer_indexes] = previous_indexes[integer_indexes]
+    # When indexes is above max index, take the max value of the array
+    indexes_above_bounds = virtual_indexes >= valid_values_count - 1
+    if indexes_above_bounds.any():
+        previous_indexes[indexes_above_bounds] = -1
+        next_indexes[indexes_above_bounds] = -1
+    # When indexes is below min index, take the min value of the array
+    indexes_below_bounds = virtual_indexes < 0
+    if indexes_below_bounds.any():
+        previous_indexes[indexes_below_bounds] = 0
+        next_indexes[indexes_below_bounds] = 0
+    if np.issubdtype(arr.dtype, np.inexact):
+        virtual_indexes_nans = np.isnan(virtual_indexes)
+        if virtual_indexes_nans.any():
+            previous_indexes[virtual_indexes_nans] = -1
+            next_indexes[virtual_indexes_nans] = -1
+    previous_indexes = previous_indexes.astype(int)
+    next_indexes = next_indexes.astype(int)
+    # --- Sorting
+    arr.partition(
+        np.concatenate(
+            ([0], np.unique(previous_indexes.ravel()),
+             np.unique(next_indexes.ravel()), [-1])),
+        axis=0)
+    if np.issubdtype(arr.dtype, np.inexact):
+        slices_having_nans = np.isnan(take(arr, -1, axis=0))
+    else:
+        slices_having_nans = None
+    # --- Get values from indexes
+    arr = arr[..., np.newaxis]
+    previous_elements = np.squeeze(
+        np.take_along_axis(arr, indices=previous_indexes[np.newaxis, ...],
+                           axis=0),
+        axis=0,
+    )
+    next_elements = np.squeeze(
+        np.take_along_axis(arr,
+                           indices=next_indexes[np.newaxis, ...],
+                           axis=0),
+        axis=0,
+    )
+    # --- Linear interpolation
+    # fuzz avoid edge cases where a upper bound would be wrongly chosen due
+    #  to a failed floating point comparison
+    # TODO find out why epsilon * 4 instead of just epsilon
+    #  (it is done like this in R and in climdex c++ impl)
```

Comment:
It looks here as well that you are specifically referring the R source code.

Reply:
This fuzz part is subject to discussion. I did an update on this in 47e8a7d4012b6c53fbae318c7ccf46ed364654aa but if necessary I guess it can be removed even if it may fix some edge cases.
This correction was not done in the former version of quantile so it could stay that way.

Reply:
Fuzz was removed, I'm resolving this.